man/tilestache-seed.1
.TH TILESTACHE-SEED 1 "Nov 10, 2010"
.SH NAME
tilestache-seed \- seed a single layer in your TileStache configuration
.SH SYNOPSIS
.B tilestache-seed
.RI [ options ] " zoom" ...
.SH DESCRIPTION
This manual page documents briefly the \fBtilestache-seed\fR command.
.PP
\fBtilestache-seed\fR seeds a single layer in your TileStache configuration.
No images are returned, but TileStache ends up with a pre-filled cache.
Bounding box is given as a pair of lat/lon coordinates, e.g. "37.788
\-122.349 37.833 \-122.246". Output is a list of tile paths as they are
created.
.br
Configuration, BBox, and Layer options are required.
.SH REQUIRED OPTIONS
.TP
\fB\-c\fR, \fB\-\-config\fR \fIfile\fR
Path to configuration file. \fBRequired\fR.
.TP
\fB-l\fR, \fB\-\-layer\fR \fIlayer\fR
Layer name from configuration. \fBRequired\fR.
.TP
\fB-b\fR, \fB\-\-bbox\fR \fIsouth\fR \fIwest\fR \fInorth\fR \fIeast\fR
Bounding box in floating point geographic coordinates. \fBRequired\fR.
.SH OPTIONS
.TP
.B \-h, \-\-help
Show summary of options.
.TP
\fB-p\fR, \fB\-\-padding\fR \fIpadding\fR
Extra margin of tiles to add around bounded area. Default value is 0 (no extra tiles).
.TP
\fB-e\fR, \fB\-\-extension\fR \fIextension\fR
Optional file type for rendered tiles. Default value is "png" for most image layers and some variety of JSON for Vector or Mapnik Grid providers.
.TP
\fB-f\fR, \fB\-\-progress-file\fR \fIfile\fR
Optional JSON progress file that gets written on each iteration, so you don't have to pay close attention.
.TP
.B \-q
Suppress chatty output, \-\-progress-file works well with this.
.TP
.B \-i, \-\-include-path
Add the following colon-separated list of paths to Python's include path (aka sys.path).
.TP
.B \-d, \-\-output-directory
Optional output directory for tiles, to override configured cache with the equivalent of:
    {"name": "Disk", "path": <output directory>, "dirs": "portable", "gzip": []}.

More information in http://tilestache.org/doc/#caches.
.TP
.B \-\-to\-mbtiles
Optional output file for tiles, will be created as an MBTiles 1.1 tileset. See http://mbtiles.org for more information.
.TP
.B \-\-from\-mbtiles
Optional input file for tiles, will be read as an MBTiles 1.1 tileset. See http://mbtiles.org for more information. Overrides \-\-extension, \-\-bbox and \-\-padding (this may change).
.TP
.B \-\-to-s3
Optional output bucket for tiles, will be populated with tiles in a standard Z/X/Y layout. Three required arguments: AWS access-key, secret, and bucket name.
.TP
.B \-\-tile\-list
Optional file of tile coordinates, a simple text list of Z/X/Y coordinates. Overrides \-\-bbox and \-\-padding.
.TP
.B \-\-error\-list
Optional file of failed tile coordinates, a simple text list of Z/X/Y coordinates. If provided, failed tiles will be logged to this file instead of stopping tilestache-seed.
.TP
.B \-\-enable\-retries
If true this will cause tilestache-seed to retry failed tile renderings up to (3) times. Default value is False.
.TP
.B \-x, \-\-ignore-cached
Re-render every tile, whether it is in the cache already or not.
.TP
.B \-\-jsonp-callback
Add a JSONP callback for tiles with a json mime-type, causing "*.js" tiles to be written to the cache wrapped in the callback function. Ignored for non-JSON tiles.
.SH SEE ALSO
.BR tilestache-render (1)
.SH AUTHOR
\fBTileStache\fR was written by Michal Migurski <mike@stamen.com>.
.PP
This manual page was written by David Paleino <dapal@debian.org>,
for the Debian project (and may be used by others).


CHANGELOG
2015-01-30: 1.50.1
- Fixed missing ttf file

2015-01-21: 1.50.0
- Added 'scale factor' option to Mapnik image provider
- Added S3 policy specification to config
- Added dictionary config to VecTiles config
- Added pixel effects to all layer configurations
- Added remote configuration to MapnikGrid to match Mapnik
- Fixed UtfGridComposite to work with latest TileStache
- Fixed Pillow dependency
- Fixed "cache lifespan" in Redis cache

2014-06-07: 1.49.11
- Fixed 'allow-origin header' getting overwritten (JesseCrocker)
- Added 'source projection' parameter to UrlTemplate provider (JesseCrocker)
- Fixed minor AttributeError when palette256 argument is None (PerryGeo)
- Added Timeout feature to Proxy and UrlTemplate providers (Juliomalegria)
- Adjusted curl examples (Peter Richardson)
- More specific error message when SpatialRef is missing (probabble)

2014-05-10: 1.49.10
- Fixed Travis build.
- Fixed import errors for case-insensitive filesystems.
- Added TileStache Vagrant machine configuration.
- Fixed some memcache testing problems.

2014-05-10: 1.49.9
- Moved everything to PyPI and fixed VERSION kerfuffle.

2013-07-02: 1.49.8
- Dropped Proxy provider srs=900913 check.
- Updated most JSON mime-types from text/json to application/json.
- Fixed requestHandler() query_string argument to match existing API docs.
- Switched order of simplify and transform in VecTiles queries to handle GEOS bug.
- Made Rummy CORS-compliant.

2013-06-23: 1.49.7
- Added VecTiles:MultiProvider to merge TopoJSON and GeoJSON output layers.

2013-06-21: 1.49.6
- Switched to ST_SimplifyPreserveTopology in VecTiles clipping to calm GEOS.

2013-06-20: 1.49.5
- Tweaked VecTiles Provider documentation.

2013-06-20: 1.49.4
- Added TopoJSON support to VecTiles Provider.

2013-06-15: 1.49.3
- Fixed runaway query bug in VecTiles Provider with looped bounds in query_columns().

2013-06-14: 1.49.2
- Fixed SQL bug in VecTiles Provider for limiting column queries.

2013-06-14: 1.49.1
- Updated magic column names in VecTiles Provider to __geometry__ and __id__ for feature ID.

2013-06-05: 1.49.0
- Added option to use S3 reduced redundancy storage for cheaper caching.
- Made Postgres connections per-request instead of long term in VecTiles Provider.

2013-05-28: 1.48.2
- Fixed a serious documentation incompatibility with TileStache.requestHandler() introduced in 1.47.0.

2013-05-27: 1.48.1
- Fixed Werkzeug dependency in setup.py.

2013-05-26: 1.48.0
- Fixed a serious documentation incompatibility with TileStache.getTile() introduced in 1.47.0.

2013-04-24: 1.47.4
- Fixed from-mbtiles/output-directory use case in tilestache-seed, which I had broken.

2013-04-22: 1.47.3
- Fixed clobbered content from TheTileLeftANote exception.

2013-04-22: 1.47.2
- Fixed getTile bugs introduced in 1.47.0.

2013-04-20: 1.47.1
- Fixed typo in setup.py.

2013-04-20: 1.47.0
- Exposed mutable HTTP status codes and headers to TileStache.getTile() and caches via TheTileLeftANote exception.
- Removed old layer.redirects implementation via TheTileIsInAnotherCastle exception.
- Improved PIL dependency in setup requirements to include Pillow, with patch from Fabian Büchler.
- Fixed coordinate generator usage in tilestache-seed with patch from Seth Fitzsimmons.
- Added Redis cache provider with patch from Matthew Perry.

2013-04-12: 1.46.10
- Prefixed most Postgis functions in VecTile provider with "ST_" for forward compatibility.

2013-04-12: 1.46.9
- Added ogr.OFTDate and ogr.OFTTime to list of acceptable field types in Vector provider.

2013-04-12: 1.46.8
- Added ogr.OFTDateTime to list of acceptable field types in Vector provider, with patch from probabble.

2013-03-27: 1.46.7
- Fixed a stupid error in VecTiles.server discarding important Postgres connection details.

2013-03-26: 1.46.6
- Improved error tolerance in VecTiles client and server.

2013-03-25: 1.46.5
- Added new simplify_until parameter to VecTiles.Provider.
- Added some invalid geometry checks to VecTiles.server.
- Added !bbox! token to VecTiles query parser.

2013-03-19: 1.46.4
- Added new zoom_data parameter to TileStache.Goodies.VecTiles Datasource.

2013-03-19: 1.46.3
- Added new simplify parameter to TileStache.Goodies.VecTiles Provider.

2013-03-14: 1.46.2
- Added new parameters to TileStache.Goodies.VecTiles Datasource and Provider.

2013-03-14: 1.46.1
- Fixed missing TileStache.Goodies.VecTiles by adding it to setup.py.

2013-03-12: 1.46.0
- Added TileStache.Goodies.VecTiles to support efficient vector data tiles for use in Mapnik.
- Fixed alpha-channel issue in Sandwich provider with patch from Lars Ahlzen.

2013-02-26: 1.45.0
- Made S3 credentials optional since Boto allows them in via environment vars or config file.
- Added path parameter to S3 cache, to use a subdirectory rather than bucket root.
- Eased bitmap size checks in Sandwich provider for retina use.
- Added "quadtile" directory layout for disk cache provider.
- Added JSON support to MBTiles provider from Aaron Cope.
- Added Travis CI tests for Vector layer.

2013-02-22: 1.44.0
- Merged Ragi Burhum's Travis continuous integration tests with ~1/3rd test coverage.
- Modified Proxy provider to return untouched original image if size and type match.
- Updated Mapnik usage to reflect recently released versions.
- Fixed Mapnik provider deadlock bug with patch from Lee Shepherd.
- Added "key prefix" argument to Memcache cache provider.

2013-01-03: 1.43.0
- Added new tilestache-list.py script to feed tilestache-seed.py --tile-list.
- Added new "use_locks" option to S3 cache provider to optionally reduce write costs.

2012-10-01: 1.42.1
- Papered over a strange appearance bug in GDAL provider with using cubicspline interpolation.

2012-09-18: 1.42.0
- Added support for tiled bitmap files in Sandwich provider layer src attribute.

2012-09-14: 1.41.0
- Added TileStache.Goodies.Providers.Cascadenik to skip Mapnik XML where possible.

2012-09-09: 1.40.1

scripts/tilestache-seed.py
#!/usr/bin/env python
"""tilestache-seed.py will warm your cache.

This script is intended to be run directly. This example seeds the area around
West Oakland (http://sta.mn/ck) in the "osm" layer, for zoom levels 12-15:

    tilestache-seed.py -c ./config.json -l osm -b 37.79 -122.35 37.83 -122.25 -e png 12 13 14 15

See `tilestache-seed.py --help` for more information.
"""

from sys import stderr, path
from os.path import realpath, dirname
from optparse import OptionParser
from urlparse import urlparse
from urllib import urlopen

try:
    from json import dump as json_dump
    from json import load as json_load
except ImportError:
    from simplejson import dump as json_dump
    from simplejson import load as json_load

#
# Most imports can be found below, after the --include-path option is known.
#

parser = OptionParser(usage="""%prog [options] [zoom...]

Seeds a single layer in your TileStache configuration - no images are returned,
but TileStache ends up with a pre-filled cache. Bounding box is given as a pair
of lat/lon coordinates, e.g. "37.788 -122.349 37.833 -122.246". Output is a list
of tile paths as they are created.

Example:

    tilestache-seed.py -b 52.55 13.28 52.46 13.51 -c tilestache.cfg -l osm 11 12 13

Protip: extract tiles from an MBTiles tileset to a directory like this:

    tilestache-seed.py --from-mbtiles filename.mbtiles --output-directory dirname

Configuration, bbox, and layer options are required; see `%prog --help` for info.""")

defaults = dict(padding=0, verbose=True, enable_retries=False, bbox=(37.777, -122.352, 37.839, -122.226))

parser.set_defaults(**defaults)

parser.add_option('-c', '--config', dest='config',
                  help='Path to configuration file, typically required.')

parser.add_option('-l', '--layer', dest='layer',
                  help='Layer name from configuration, typically required.')

parser.add_option('-b', '--bbox', dest='bbox',
                  help='Bounding box in floating point geographic coordinates: south west north east. Default value is %.7f, %.7f, %.7f, %.7f.' % defaults['bbox'],
                  type='float', nargs=4)

parser.add_option('-p', '--padding', dest='padding',
                  help='Extra margin of tiles to add around bounded area. Default value is %s (no extra tiles).' % repr(defaults['padding']),
                  type='int')

parser.add_option('-e', '--extension', dest='extension',
                  help='Optional file type for rendered tiles. Default value is "png" for most image layers and some variety of JSON for Vector or Mapnik Grid providers.')

parser.add_option('-f', '--progress-file', dest='progressfile',
                  help="Optional JSON progress file that gets written on each iteration, so you don't have to pay close attention.")

parser.add_option('-q', action='store_false', dest='verbose',
                  help='Suppress chatty output, --progress-file works well with this.')

parser.add_option('-i', '--include-path', dest='include_paths',
                  help="Add the following colon-separated list of paths to Python's include path (aka sys.path)")

parser.add_option('-d', '--output-directory', dest='outputdirectory',
                  help='Optional output directory for tiles, to override configured cache with the equivalent of: {"name": "Disk", "path": <output directory>, "dirs": "portable", "gzip": []}. More information in http://tilestache.org/doc/#caches.')

parser.add_option('--to-mbtiles', dest='mbtiles_output',
                  help='Optional output file for tiles, will be created as an MBTiles 1.1 tileset. See http://mbtiles.org for more information.')

parser.add_option('--to-s3', dest='s3_output',
                  help='Optional output bucket for tiles, will be populated with tiles in a standard Z/X/Y layout. Three required arguments: AWS access-key, secret, and bucket name.',
                  nargs=3)

parser.add_option('--from-mbtiles', dest='mbtiles_input',
                  help='Optional input file for tiles, will be read as an MBTiles 1.1 tileset. See http://mbtiles.org for more information. Overrides --extension, --bbox and --padding (this may change).')

parser.add_option('--tile-list', dest='tile_list',
                  help='Optional file of tile coordinates, a simple text list of Z/X/Y coordinates. Overrides --bbox and --padding.')

parser.add_option('--error-list', dest='error_list',
                  help='Optional file of failed tile coordinates, a simple text list of Z/X/Y coordinates. If provided, failed tiles will be logged to this file instead of stopping tilestache-seed.')

parser.add_option('--enable-retries', dest='enable_retries',
                  help='If true this will cause tilestache-seed to retry failed tile renderings up to (3) times. Default value is %s.' % repr(defaults['enable_retries']),
                  action='store_true')

parser.add_option('-x', '--ignore-cached', action='store_true', dest='ignore_cached',
                  help='Re-render every tile, whether it is in the cache already or not.')

parser.add_option('--jsonp-callback', dest='callback',
                  help='Add a JSONP callback for tiles with a json mime-type, causing "*.js" tiles to be written to the cache wrapped in the callback function. Ignored for non-JSON tiles.')

def generateCoordinates(ul, lr, zooms, padding):
    """ Generate a stream of (offset, count, coordinate) tuples for seeding.
    
        Flood-fill coordinates based on two corners, a list of zooms and padding.
    """
    # start with a simple total of all the coordinates we will need.
    count = 0
    
    for zoom in zooms:
        ul_ = ul.zoomTo(zoom).container().left(padding).up(padding)
        lr_ = lr.zoomTo(zoom).container().right(padding).down(padding)
        
        rows = lr_.row + 1 - ul_.row
        cols = lr_.column + 1 - ul_.column
        
        count += int(rows * cols)

    # now generate the actual coordinates.
    # offset starts at zero
    offset = 0
    
    for zoom in zooms:
        ul_ = ul.zoomTo(zoom).container().left(padding).up(padding)
        lr_ = lr.zoomTo(zoom).container().right(padding).down(padding)

        for row in xrange(int(ul_.row), int(lr_.row + 1)):
            for column in xrange(int(ul_.column), int(lr_.column + 1)):
                coord = Coordinate(row, column, zoom)
                
                yield (offset, count, coord)
                
                offset += 1

def listCoordinates(filename):
    """ Generate a stream of (offset, count, coordinate) tuples for seeding.
    
        Read coordinates from a file with one Z/X/Y coordinate per line.
    """
    coords = (line.strip().split('/') for line in open(filename, 'r'))
    coords = (map(int, (row, column, zoom)) for (zoom, column, row) in coords)
    coords = [Coordinate(*args) for args in coords]
    
    count = len(coords)
    
    for (offset, coord) in enumerate(coords):
        yield (offset, count, coord)

def tilesetCoordinates(filename):
    """ Generate a stream of (offset, count, coordinate) tuples for seeding.
    
        Read coordinates from an MBTiles tileset filename.
    """
    coords = MBTiles.list_tiles(filename)
    count = len(coords)
    
    for (offset, coord) in enumerate(coords):
        yield (offset, count, coord)

def parseConfigfile(configpath):
    """ Parse a configuration file and return a raw dictionary and dirpath.
    
        Return value can be passed to TileStache.Config.buildConfiguration().
    """
    config_dict = json_load(urlopen(configpath))
    
    scheme, host, path, p, q, f = urlparse(configpath)
    
    if scheme == '':
        scheme = 'file'
        path = realpath(path)
    
    dirpath = '%s://%s%s' % (scheme, host, dirname(path).rstrip('/') + '/')
    
    return config_dict, dirpath

if __name__ == '__main__':
    options, zooms = parser.parse_args()

    if options.include_paths:
        for p in options.include_paths.split(':'):
            path.insert(0, p)

    from TileStache import getTile, Config
    from TileStache.Core import KnownUnknown
    from TileStache.Config import buildConfiguration
    from TileStache import MBTiles
    import TileStache
    
    from ModestMaps.Core import Coordinate
    from ModestMaps.Geo import Location

    try:
        # determine if we have enough information to prep a config and layer
        
        has_fake_destination = bool(options.outputdirectory or options.mbtiles_output)
        has_fake_source = bool(options.mbtiles_input)
        
        if has_fake_destination and has_fake_source:
            config_dict, config_dirpath = dict(layers={}), '' # parseConfigfile(options.config)
            layer_dict = dict()
            
            config_dict['cache'] = dict(name='test')
            config_dict['layers'][options.layer or 'tiles-layer'] = layer_dict
        
        elif options.config is None:
            raise KnownUnknown('Missing required configuration (--config) parameter.')
        
        elif options.layer is None:
            raise KnownUnknown('Missing required layer (--layer) parameter.')
    
        else:
            config_dict, config_dirpath = parseConfigfile(options.config)
            
            if options.layer not in config_dict['layers']:
                raise KnownUnknown('"%s" is not a layer I know about. Here are some that I do know about: %s.' % (options.layer, ', '.join(sorted(config_dict['layers'].keys()))))
            
            layer_dict = config_dict['layers'][options.layer]
            layer_dict['write_cache'] = True # Override to make seeding guaranteed useful.
        
        # override parts of the config and layer if needed
        
        extension = options.extension

        if options.mbtiles_input:
            layer_dict['provider'] = dict(name='mbtiles', tileset=options.mbtiles_input)
            n, t, v, d, format, b = MBTiles.tileset_info(options.mbtiles_input)
            extension = format or extension
        
        # determine or guess an appropriate tile extension
        
        if extension is None:
            provider_name = layer_dict['provider'].get('name', '').lower()
            
            if provider_name == 'mapnik grid':
                extension = 'json'
            elif provider_name == 'vector':
                extension = 'geojson'
            else:
                extension = 'png'
        
        # override parts of the config and layer if needed
        
        tiers = []
        
        if options.mbtiles_output:
            tiers.append({'class': 'TileStache.MBTiles:Cache',
                          'kwargs': dict(filename=options.mbtiles_output,
                                         format=extension,
                                         name=options.layer)})
        
        if options.outputdirectory:
            tiers.append(dict(name='disk', path=options.outputdirectory,
                              dirs='portable', gzip=[]))

        if options.s3_output:
            access, secret, bucket = options.s3_output
            tiers.append(dict(name='S3', bucket=bucket,
                              access=access, secret=secret))
        
        if len(tiers) > 1:
            config_dict['cache'] = dict(name='multi', tiers=tiers)
        elif len(tiers) == 1:
            config_dict['cache'] = tiers[0]
        else:
            # Leave config_dict['cache'] as-is
            pass
        
        # create a real config object
        
        config = buildConfiguration(config_dict, config_dirpath)
        layer = config.layers[options.layer or 'tiles-layer']
        
        # do the actual work
        
        lat1, lon1, lat2, lon2 = options.bbox
        south, west = min(lat1, lat2), min(lon1, lon2)
        north, east = max(lat1, lat2), max(lon1, lon2)

        if not (-90.0 < south < 90.0) or not (-90.0 < north < 90.0):
            raise KnownUnknown(
                'Latitude must be a value between -90 and 90 '
                '(Hint: Maybe you did long/lat instead of lat/long?).'
            )
        if not (-180.0 < west < 180.0) or not (-180.0 < east < 180.0):
            raise KnownUnknown(
                'Longitude must be a value between -180 and 180.'
            )

        northwest = Location(north, west)
        southeast = Location(south, east)

        ul = layer.projection.locationCoordinate(northwest)
        lr = layer.projection.locationCoordinate(southeast)

        for (i, zoom) in enumerate(zooms):
            if not zoom.isdigit():
                raise KnownUnknown('"%s" is not a valid numeric zoom level.' % zoom)

            zooms[i] = int(zoom)
        
        if options.padding < 0:
            raise KnownUnknown('A negative padding will not work.')

        padding = options.padding
        tile_list = options.tile_list
        error_list = options.error_list

    except KnownUnknown, e:
        parser.error(str(e))

    if tile_list:
        coordinates = listCoordinates(tile_list)
    elif options.mbtiles_input:
        coordinates = tilesetCoordinates(options.mbtiles_input)
    else:
        coordinates = generateCoordinates(ul, lr, zooms, padding)
    
    for (offset, count, coord) in coordinates:
        path = '%s/%d/%d/%d.%s' % (layer.name(), coord.zoom, coord.column, coord.row, extension)

        progress = {"tile": path,
                    "offset": offset + 1,
                    "total": count}

        #
        # Fetch a tile.
        #
        
        attempts = options.enable_retries and 3 or 1
        rendered = False
        
        while not rendered:
            if options.verbose:
                print >> stderr, '%(offset)d of %(total)d...' % progress,
    
            try:
                mimetype, content = getTile(layer, coord, extension, options.ignore_cached)
                
                if mimetype and 'json' in mimetype and options.callback:
                    js_path = '%s/%d/%d/%d.js' % (layer.name(), coord.zoom, coord.column, coord.row)
                    js_body = '%s(%s);' % (options.callback, content)
                    js_size = len(js_body) / 1024
                    
                    layer.config.cache.save(js_body, layer, coord, 'JS')
                    print >> stderr, '%s (%dKB)' % (js_path, js_size),
            
                elif options.callback:
                    print >> stderr, '(callback ignored)',
            
            except:
                #
                # Something went wrong: try again? Log the error?
                #
                attempts -= 1

                if options.verbose:
                    print >> stderr, 'Failed %s, will try %s more.' % (progress['tile'], ['no', 'once', 'twice'][attempts])
                
                if attempts == 0:
                    if not error_list:
                        raise
                    
                    fp = open(error_list, 'a')
                    fp.write('%(zoom)d/%(column)d/%(row)d\n' % coord.__dict__)
                    fp.close()
                    break
            
            else:
                #
                # Successfully got the tile.
                #
                rendered = True
                progress['size'] = '%dKB' % (len(content) / 1024)
        
                if options.verbose:
                    print >> stderr, '%(tile)s (%(size)s)' % progress
                
        if options.progressfile:
            fp = open(options.progressfile, 'w')
            json_dump(progress, fp)
            fp.close()
